Objects/regions.c

```diff
@@ -94,9 +120,299 @@ struct regionmetadata {
+    }
+
+    while (self) {
+        if (self == other) {
+            return true;
+        }
+        self = regionmetadata_get_parent(self);
+    }
+    return false;
+}
+
+
+// This implementation merges `self` into `other`. Merging is not allowed
+// to break external uniqueness. It's therefore not allowed if both regions
+// to have a parent. Except cases, where one region has the other region as
+// it's parent.
+//
+// This function expects `self` to be a valid object.
+__attribute__((unused))
+static PyObject* regionmetadata_merge(regionmetadata* self, Py_region_ptr_t other) {
+    assert(HAS_METADATA(self) && "The immutable and local region can't be merged into another region");
+    assert(REGION_PTR_CAST(self) == regionmetadata_get_root(self) && "Sanity Check");
+
+    // If `other` is the parent of `self` we can merge it. We unset the the
+    // parent which will also update the rc and other counts.
+    regionmetadata* self_parent = regionmetadata_get_parent(self);
+    if (REGION_PTR_CAST(self_parent) == other) {
+        assert(HAS_METADATA(self_parent) && "The immutable and local region can never have children");
+
+        regionmetadata_set_parent(self, NULL);
+        self_parent = NULL;
+    }
+
+    // If only `self` has a parent we can make `other` the child and
+    // remove the parent from `self`. The merged region will then again
+    // have the correct parent.
+    regionmetadata* other_parent = regionmetadata_get_parent(self);
+    if (self_parent && HAS_METADATA(other) && other_parent == NULL) {
+        // Make sure we don't create any cycles
+        if (regionmetadata_has_ancestor(self_parent, REGION_DATA_CAST(other))) {
+            throw_region_error(self->bridge, REGION_DATA_CAST(other)->bridge,
+                        "Merging these regions would create a cycle", NULL);
+            return NULL;
+        }
+
+        regionmetadata_set_parent(REGION_DATA_CAST(other), self_parent);
+        regionmetadata_set_parent(self, NULL);
+        self_parent = NULL;
+    }
+
+    // If `self` still has a parent we can't merge it into `other`
+    if (self_parent != NULL) {
+        PyObject* other_node = NULL;
+        if (HAS_METADATA(other))  {
+            other_node = _PyObject_CAST(REGION_DATA_CAST(other)->bridge);
+        }
+        throw_region_error(self->bridge, other_node,
+                        "Unable to merge regions", NULL);
+        return NULL;
+    }
+
+    regionmetadata_inc_rc(other);
+
+    // Move LRC and OSC into the root.
+    if (HAS_METADATA(other)) {
+        // Move information into the merge root
+        regionmetadata* other_data = REGION_DATA_CAST(other);
+        other_data->lrc += self->lrc;
+        other_data->osc += self->osc;
+        other_data->is_open |= self->is_open;
+        // remove information from self
+        self->lrc = 0;
+        self->osc = 0;
+        self->is_open = false;
+    }
+
+    self->parent = Py_region_ptr_with_tags(other | Py_METADATA_MERGE_TAG);
```

Comment:
Maybe a nice macro or function called lazy_merge or something would be more readable than this? I don't know what the best name is but it would be good to raise the level of abstraction above the implementation IMO.